cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/co/cask/cdap/etl/spark/SparkPipelineRunner.java

```diff
@@ -329,6 +342,42 @@ public void runPipeline(PipelinePhase pipelinePhase, String sourcePluginType,
 
       emittedRecords.put(stageName, emittedBuilder.build());
     }
+
+    Collection<Future> sinkFutures = new ArrayList<>(sinkRunnables.size());
+    ExecutorService executorService = Executors.newFixedThreadPool(sinkRunnables.size(), new ThreadFactoryBuilder()
+      .setNameFormat("pipeline-sink-task")
+      .build());
+    for (Runnable runnable : sinkRunnables) {
+      sinkFutures.add(executorService.submit(runnable));
+    }
+
+    Throwable error = null;
+    Iterator<Future> futureIter = sinkFutures.iterator();
+    while (futureIter.hasNext()) {
+      Future future = futureIter.next();
+      try {
+        future.get();
+      } catch (ExecutionException | InterruptedException e) {
+        Throwable cause = e;
+        if (e instanceof ExecutionException) {
+          cause = e.getCause();
+        }
+        if (error == null) {
+          error = cause;
+        } else {
+          error.addSuppressed(cause);
+        }
+        LOG.error("A spark job failed, cancelling all other jobs.");
+        while (futureIter.hasNext()) {
+          Future otherFuture = futureIter.next();
+          otherFuture.cancel(true);
```

Comment:
If you call `shutdownNow` on the executor service, the effect is the same as this loop. Probably don't need this loop.